dbms/src/DataTypes/DataTypeDecimalBase.h

```diff
@@ -0,0 +1,225 @@
+        return getScaleMultiplier(scale_delta);
+    }
+
+    template <typename U>
+    T scaleFactorFor(const DataTypeNumber<U> & , bool is_multiply_or_divisor) const
+    {
+        if (is_multiply_or_divisor)
+            return 1;
+        return getScaleMultiplier();
+    }
+
+    static T getScaleMultiplier(UInt32 scale);
+
+protected:
+    const UInt32 precision;
+    const UInt32 scale;
+};
+
+
+template <typename T, typename U, template <typename> typename DecimalType>
+typename std::enable_if_t<(sizeof(T) >= sizeof(U)), DecimalType<T>>
+decimalResultType(const DecimalType<T> & tx, const DecimalType<U> & ty, bool is_multiply, bool is_divide)
+{
+    UInt32 scale = (tx.getScale() > ty.getScale() ? tx.getScale() : ty.getScale());
+    if (is_multiply)
+        scale = tx.getScale() + ty.getScale();
+    else if (is_divide)
+        scale = tx.getScale();
+    return DecimalType<T>(maxDecimalPrecision<T>(), scale);
+}
+
+template <typename T, typename U, template <typename> typename DecimalType>
+typename std::enable_if_t<(sizeof(T) < sizeof(U)), const DecimalType<U>>
+decimalResultType(const DecimalType<T> & tx, const DecimalType<U> & ty, bool is_multiply, bool is_divide)
+{
+    UInt32 scale = (tx.getScale() > ty.getScale() ? tx.getScale() : ty.getScale());
+    if (is_multiply)
+        scale = tx.getScale() * ty.getScale();
+    else if (is_divide)
+        scale = tx.getScale();
+    return DecimalType<U>(maxDecimalPrecision<U>(), scale);
+}
+
+template <typename T, typename U, template <typename> typename DecimalType>
+const DecimalType<T> decimalResultType(const DecimalType<T> & tx, const DataTypeNumber<U> &, bool, bool)
+{
+    return DecimalType<T>(maxDecimalPrecision<T>(), tx.getScale());
+}
+
+template <typename T, typename U, template <typename> typename DecimalType>
+const DecimalType<U> decimalResultType(const DataTypeNumber<T> &, const DecimalType<U> & ty, bool, bool)
+{
+    return DecimalType<U>(maxDecimalPrecision<U>(), ty.getScale());
+}
+
+template <template <typename> typename DecimalType>
+DataTypePtr createDecimal(UInt64 precision_value, UInt64 scale_value)
+{
+    if (precision_value < minDecimalPrecision() || precision_value > maxDecimalPrecision<Decimal128>())
+        throw Exception("Wrong precision", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+
+    if (static_cast<UInt64>(scale_value) > precision_value)
+        throw Exception("Negative scales and scales larger than precision are not supported", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+
+    if (precision_value <= maxDecimalPrecision<Decimal32>())
+        return std::make_shared<DecimalType<Decimal32>>(precision_value, scale_value);
+    else if (precision_value <= maxDecimalPrecision<Decimal64>())
+        return std::make_shared<DecimalType<Decimal64>>(precision_value, scale_value);
+    return std::make_shared<DecimalType<Decimal128>>(precision_value, scale_value);
+}
+
+extern template class DataTypeDecimalBase<Decimal32>;
```

Comment:
Am I right it's the first place in our code base with extern template? Could we avoid them?

Reply:
sure, let me remove it

Reply:
done